In [1]:
! pip install deepchem

In [13]:
import deepchem as dc
from deepchem.feat.molecule_featurizers import MolGraphConvFeaturizer

In [30]:
feat = MolGraphConvFeaturizer(use_edges=True)

toxcast_tasks, toxcast_datasets, transformers = dc.molnet.load_toxcast(featurizer=feat)
train_dataset, valid_dataset, test_dataset = toxcast_datasets

print(len(train_dataset), len(valid_dataset), len(test_dataset))

Failed to featurize datapoint 47, [Ca+2].[Cl-].[Cl-]. Appending empty array
Exception message: tuple index out of range
Failed to featurize datapoint 93, [Cd+2].[Cl-].[Cl-]. Appending empty array
Exception message: tuple index out of range
Failed to featurize datapoint 220, O.O.[Ba+2].[Cl-].[Cl-]. Appending empty array
Exception message: tuple index out of range
Failed to featurize datapoint 1039, None. Appending empty array
Exception message: Python argument types in
    rdkit.Chem.rdmolfiles.CanonicalRankAtoms(NoneType)
did not match C++ signature:
    CanonicalRankAtoms(class RDKit::ROMol mol, bool breakTies=True, bool includeChirality=True, bool includeIsotopes=True)
Failed to featurize datapoint 1789, None. Appending empty array
Exception message: Python argument types in
    rdkit.Chem.rdmolfiles.CanonicalRankAtoms(NoneType)
did not match C++ signature:
    CanonicalRankAtoms(class RDKit::ROMol mol, bool breakTies=True, bool includeChirality=True, bool includeIsotopes=True)
Faile

6852 857 857


In [31]:
toxcast_tasks_one, toxcast_datasets_one, transformers_one = dc.molnet.load_toxcast(featurizer="GraphConv")
train_dataset_one, valid_dataset_one, test_dataset_one = toxcast_datasets_one

print(len(train_dataset_one), len(valid_dataset_one), len(test_dataset_one))

Failed to featurize datapoint 1039, None. Appending empty array
Exception message: Python argument types in
    rdkit.Chem.rdmolfiles.CanonicalRankAtoms(NoneType)
did not match C++ signature:
    CanonicalRankAtoms(class RDKit::ROMol mol, bool breakTies=True, bool includeChirality=True, bool includeIsotopes=True)
Failed to featurize datapoint 1789, None. Appending empty array
Exception message: Python argument types in
    rdkit.Chem.rdmolfiles.CanonicalRankAtoms(NoneType)
did not match C++ signature:
    CanonicalRankAtoms(class RDKit::ROMol mol, bool breakTies=True, bool includeChirality=True, bool includeIsotopes=True)
Failed to featurize datapoint 1881, None. Appending empty array
Exception message: Python argument types in
    rdkit.Chem.rdmolfiles.CanonicalRankAtoms(NoneType)
did not match C++ signature:
    CanonicalRankAtoms(class RDKit::ROMol mol, bool breakTies=True, bool includeChirality=True, bool includeIsotopes=True)
Failed to featurize datapoint 2456, None. Appending emp

6861 858 858


In [22]:
! pip install dgl dgllife

     -------------------------------------- 226.1/226.1 kB 3.5 MB/s eta 0:00:00
     ---------------------------------------- 1.6/1.6 MB 14.4 MB/s eta 0:00:00
  Using cached future-0.18.2.tar.gz (829 kB)
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
     ------------------------------------- 200.5/200.5 kB 11.9 MB/s eta 0:00:00
  Created wheel for future: filename=future-0.18.2-py3-none-any.whl size=491086 sha256=942a7f14da26b793291430e024cfd24a81724c93d06a8a3a47f9c0800e7076ae
  Stored in directory: c:\users\user\appdata\local\pip\cache\wheels\dc\16\09\eb08b4e34e6b638f113d2018cf0b22de1d8dca22a3a71873f7
Successfully built future


In [32]:
from deepchem.models import GATModel, GCNModel, GraphConvModel

gcn_model_one = GraphConvModel(n_tasks=len(toxcast_tasks_one), mode="classification", dropout=0.1, batch_size=32, learning_rate=0.001)
gat_model = GATModel(mode="classification", dropout=0.1, n_tasks=len(toxcast_tasks), learning_rate=0.001, batch_size=32)
gcn_model = GCNModel(n_tasks=len(toxcast_tasks), mode="classification", dropout=0.1, batch_size=32, learning_rate=0.001)

In [33]:
gcn_model.fit(train_dataset, nb_epoch = 50)

0.21976844787597657

In [39]:
import numpy as np

metric = dc.metrics.Metric(dc.metrics.balanced_accuracy_score, np.mean, mode="classification")
train_scores = gcn_model.evaluate(train_dataset, [metric], transformers)
test_scores = gcn_model.evaluate(test_dataset, [metric], transformers)

print(f"Training Balanced Accuracy: {train_scores}:.3f")
print(f"Testing Balanced Accuracy: {test_scores}:.3f")

c:\Users\user\OneDrive\Desktop\GitHub\small-molecules\env\lib\site-packages\sklearn\metrics\_classification.py:2184: UserWarning: y_pred contains classes not in y_true
  warnings.warn("y_pred contains classes not in y_true")
c:\Users\user\OneDrive\Desktop\GitHub\small-molecules\env\lib\site-packages\sklearn\metrics\_classification.py:2184: UserWarning: y_pred contains classes not in y_true
  warnings.warn("y_pred contains classes not in y_true")
c:\Users\user\OneDrive\Desktop\GitHub\small-molecules\env\lib\site-packages\sklearn\metrics\_classification.py:2184: UserWarning: y_pred contains classes not in y_true
  warnings.warn("y_pred contains classes not in y_true")
c:\Users\user\OneDrive\Desktop\GitHub\small-molecules\env\lib\site-packages\sklearn\metrics\_classification.py:2184: UserWarning: y_pred contains classes not in y_true
  warnings.warn("y_pred contains classes not in y_true")
c:\Users\user\OneDrive\Desktop\GitHub\small-molecules\env\lib\site-packages\sklearn\metrics\_classif

Training Balanced Accuracy: {'mean-balanced_accuracy_score': 0.8169906374246726}:.3f
Testing Balanced Accuracy: {'mean-balanced_accuracy_score': 0.601552018776808}:.3f


c:\Users\user\OneDrive\Desktop\GitHub\small-molecules\env\lib\site-packages\sklearn\metrics\_classification.py:2184: UserWarning: y_pred contains classes not in y_true
  warnings.warn("y_pred contains classes not in y_true")
c:\Users\user\OneDrive\Desktop\GitHub\small-molecules\env\lib\site-packages\sklearn\metrics\_classification.py:2184: UserWarning: y_pred contains classes not in y_true
  warnings.warn("y_pred contains classes not in y_true")


In [34]:
gat_model.fit(train_dataset, nb_epoch=50)

0.2801251220703125

In [40]:
train_scores = gat_model.evaluate(train_dataset, [metric], transformers)
test_scores = gat_model.evaluate(test_dataset, [metric], transformers)

print(f"Training Balanced Accuracy: {train_scores}:.3f")
print(f"Testing Balanced Accuracy: {test_scores}:.3f")

c:\Users\user\OneDrive\Desktop\GitHub\small-molecules\env\lib\site-packages\sklearn\metrics\_classification.py:2184: UserWarning: y_pred contains classes not in y_true
  warnings.warn("y_pred contains classes not in y_true")
c:\Users\user\OneDrive\Desktop\GitHub\small-molecules\env\lib\site-packages\sklearn\metrics\_classification.py:2184: UserWarning: y_pred contains classes not in y_true
  warnings.warn("y_pred contains classes not in y_true")
c:\Users\user\OneDrive\Desktop\GitHub\small-molecules\env\lib\site-packages\sklearn\metrics\_classification.py:2184: UserWarning: y_pred contains classes not in y_true
  warnings.warn("y_pred contains classes not in y_true")
c:\Users\user\OneDrive\Desktop\GitHub\small-molecules\env\lib\site-packages\sklearn\metrics\_classification.py:2184: UserWarning: y_pred contains classes not in y_true
  warnings.warn("y_pred contains classes not in y_true")
c:\Users\user\OneDrive\Desktop\GitHub\small-molecules\env\lib\site-packages\sklearn\metrics\_classif

Training Balanced Accuracy: {'mean-balanced_accuracy_score': 0.774433799765263}:.3f
Testing Balanced Accuracy: {'mean-balanced_accuracy_score': 0.601090328562027}:.3f


c:\Users\user\OneDrive\Desktop\GitHub\small-molecules\env\lib\site-packages\sklearn\metrics\_classification.py:2184: UserWarning: y_pred contains classes not in y_true
  warnings.warn("y_pred contains classes not in y_true")
c:\Users\user\OneDrive\Desktop\GitHub\small-molecules\env\lib\site-packages\sklearn\metrics\_classification.py:2184: UserWarning: y_pred contains classes not in y_true
  warnings.warn("y_pred contains classes not in y_true")


In [36]:
gcn_model_one.fit(train_dataset_one, nb_epoch=50)

0.20547828674316407

In [44]:
train_scores = gcn_model_one.evaluate(train_dataset_one, [metric], transformers_one)
test_scores = gcn_model_one.evaluate(test_dataset_one, [metric], transformers_one)

print(f"Training Balanced Accuracy: {train_scores}")
print(f"Testing Balanced Accuracy: {test_scores}")

c:\Users\user\OneDrive\Desktop\GitHub\small-molecules\env\lib\site-packages\sklearn\metrics\_classification.py:2184: UserWarning: y_pred contains classes not in y_true
  warnings.warn("y_pred contains classes not in y_true")
c:\Users\user\OneDrive\Desktop\GitHub\small-molecules\env\lib\site-packages\sklearn\metrics\_classification.py:2184: UserWarning: y_pred contains classes not in y_true
  warnings.warn("y_pred contains classes not in y_true")
c:\Users\user\OneDrive\Desktop\GitHub\small-molecules\env\lib\site-packages\sklearn\metrics\_classification.py:2184: UserWarning: y_pred contains classes not in y_true
  warnings.warn("y_pred contains classes not in y_true")
c:\Users\user\OneDrive\Desktop\GitHub\small-molecules\env\lib\site-packages\sklearn\metrics\_classification.py:2184: UserWarning: y_pred contains classes not in y_true
  warnings.warn("y_pred contains classes not in y_true")
c:\Users\user\OneDrive\Desktop\GitHub\small-molecules\env\lib\site-packages\sklearn\metrics\_classif

Training Balanced Accuracy: {'mean-balanced_accuracy_score': 0.8431794694110958}
Testing Balanced Accuracy: {'mean-balanced_accuracy_score': 0.5897727212107009}


c:\Users\user\OneDrive\Desktop\GitHub\small-molecules\env\lib\site-packages\sklearn\metrics\_classification.py:2184: UserWarning: y_pred contains classes not in y_true
  warnings.warn("y_pred contains classes not in y_true")
c:\Users\user\OneDrive\Desktop\GitHub\small-molecules\env\lib\site-packages\sklearn\metrics\_classification.py:2184: UserWarning: y_pred contains classes not in y_true
  warnings.warn("y_pred contains classes not in y_true")


Observations
-----------

## Baseline

We'll use the GCN Model, with MolGraphConvFeaturizer featurization as the baseline estimator, 
hence no model should do worse than it.

- Training Balanced Accuracy: 84.320%

- Testing Balanced Accuracy: 60%

## Preliminary model

Our preliminary model is a GraphConvModel with GraphConvMol featurization

- Training Balanced Accuracy: 81.699%

- Testing Balanced Accuracy: 60.155%

In [ ]:
## Hyperparameter optimization

def model_bulder(**model_params):
    model = GraphConvModel(**model_params)
    return model

from deepchem.hyper import HyperparamOpt

params = {}

optimizer = HyperparamOpt(model_bulder)
best_model, best_hyperparams, all_results =   optimizer.hyperparam_search(params, train_dataset, test_dataset, metric)

best_hyperparams